doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb

```diff
@@ -118,18 +114,22 @@
     "\n",
     "# create the networks\n",
     "net_no_reg = NeuralNetwork(layers)\n",
+    "net_no_reg.put(\"seed\", 10)\n",
     "net_no_reg.quick_connect()\n",
     "net_no_reg.initialize_neural_network()\n",
     "\n",
     "net_l2 = NeuralNetwork(layers)\n",
+    "net_l2.put(\"seed\", 10)\n",
     "net_l2.quick_connect()\n",
     "net_l2.initialize_neural_network()\n",
     "\n",
     "net_l1 = NeuralNetwork(layers)\n",
```

Comment:
if you touch bits like this, and the ctors are already available through factories, it would be great if you could replace them, e.g. https://github.com/shogun-toolbox/shogun/blob/develop/examples/meta/src/neural_nets/feedforward_net_classification.sg.in

Of course, only if this works out of the box (e.g. when there are examples to copy paste it from).and you have the nerves ;)

Reply:
Ah cool, I thought it wasn't working properly with factories, because the old API is still here. I'll have a look. Now the CI is finally actually running the job.

Reply:
actually this is not being run by the interfaces...

Reply:
@karlnapf How do you think we should expose `DeepAutoencoder::pre_train(std::shared_ptr<Features>)`?

Reply:
and DeepAutoencoder::std::shared_ptr<DenseFeatures<float64_t>> data>)?

Reply:
@vigsterkr thoughts^

Reply:
ain't got no DenseFeatures in SWIG :(

Reply:
this is one of the cases where it is a bit of work. We need to provide an abstraction for such operations in the base api. So far, we managed to work around this, via something like
```
nn.put('pre_train_features', f)
nn.run('pre_train') # throws error if above wasnt done by the user
```
This is also done for things like labels (runtime error if not set), kernels, and more specific situations. I am not the biggest fan (especially for labels, which imo should be passed to `::train`, but that is another discussion), but there *will* be methods that are specific to particular classes and wont be in the base api.

Another solution could be a `watch_method` with a single sgobject parameter (see our discussion about adding one with an index), which is then dynamically casted to the desired class. But extending the base API in this matter is also at risk of making things complicated.

The root problem really is the design of the NN class here. Under Shogun's paradigm, it is bad style of relying on such mechanics. So a re-design of the class would be best.